doc/developer/design/20231204_query_lifecycle_events_logging.md

```diff
@@ -0,0 +1,149 @@
+| `sources_became_ready`              | When all transitive source dependencies advanced to the timestamp. If there are none, same as `execution_began`             |
+| `compute_dependencies_became_ready` | When all compute dependencies (MVs and Indexes) advanced to the timestamp. If there are none, same as `execution_began`.    |
+| `execution_finished`                | Corresponds to today's `mz_statement_execution_history.finished_at`                                                         |
+| `last_row_was_returned`             | The time at which all rows were sent and we were ready to receive new statements from its session. |
+
+When each event is generated, the coordinator will insert it into a
+buffer which is periodically dumped to storage, similarly to the
+statement logging events that exist today. Since the amount of data is
+proportional to the number of statements, we don't envision that this
+will result in new scalability concerns beyond those which already
+exist for the statement log. 
+
+In the following subsections we will explain how to collect each of
+the pieces of data.
+
+### `execution_began`
+
+We already collect this.
+
+### `sources_became_ready` / `compute_dependencies_became_ready`
+
+Introduce a mechanism for the sub-controllers (Storage and Compute
+controllers) to return all frontier updates to the overall
+Controller. When installing a query with the Controller, the
+Coordinator will also inform it of one or more "watch sets" of
+IDs. Concretely, the Controller will gain a function like:
+
+``` rust
+fn add_watch_set(&mut self, id: Box<Any>, global_ids: Vec<GlobalId>, ts: Timestamp);
+```
+
+The Controller will then be expected to return responses of the form:
+
+``` rust
+WatchSetFulfilled {
+    id: Box<Any>,
+}
+```
+
+when the write frontier for all the global IDs in the set has passed
+`ts`. The coordinator will then emit the corresponding events.
+
+### `execution_finished`
+
+We already collect this.
+
+### `last_row_was_returned`
+
+We would need to start sending the statement logging id to the client
+(pgwire/sql) layer along with all execute responses. Currently, we
+only send it with the execute response variants for which
+`finished_at` is determined in that layer.
+
+Whenever we call `flush` in the pgwire layer (or its equivalent in the
+HTTP layer), we can log this event for all the statement logging IDs
+for which we have enqueued responses since the last call to `flush`.
+
+## Open questions
+
+* Have we gotten the granularity of events right? I'm slightly
+  suspicious of the split between "sources" and "compute
+  dependencies". If a computation on one cluster depends on an MV
+  being computed on another cluster, and so on, might we want to
+  record an event when _each cluster_ in that chain has finished its
+  work? On the other hand, this would start to duplicate the work for
+  the Workflows Graph.
```

Comment:
Me too. Arguably there should be an "input became ready" event for each input in the transitive dependency graph.  That might make more sense in theory than in practice.

Reply:
Curious to hear @frankmcsherry 's take given that the idea for this table ultimately originates with him.

Reply:
I think there was a take that a coarse granularity made some sense, because it was helpful to break things into "waiting on data", "waiting on compute", "waiting on serving". Each of them could be unpacked for sure! Unpack "waiting on data" to each of the inputs, unpack "waiting on compute" to the stages of indexes that come on line, etc.

Does that check out?

Reply:
It does! But it is a little weird to me (but plausibly fine) that "coarse" for compute means "immediate dependencies" while "coarse" for storage means "transitive dependencies". Mentioned originally here: https://github.com/MaterializeInc/materialize/pull/23670/files/07ffec3d88b940b6ead74542a0faeb5a6b97f079#r1414857905.

Reply:
I think "coarse" should mean transitive dependencies for both! It's just that for compute, the two notions are the same (since compute dependencies will always become ready in topological order)